Add check that there are some resource entries

nhelfman opened the pull request:
Test resource_initiator_types doesn't check that any entries have returned from resrouce timing.
The test currently doesn't return stable number of entries but we need to at least ensure there are some entries.

Commit:
Add assert that there are some resource entries

Test resource_initiator_types doesn't check that any entries have returned from resrouce timing.
The test currently doesn't return stable number of entries but we need to at least ensure there are some entries.

nhelfman commented on resource-timing/resource_initiator_types.html:
```diff
@@ -135,6 +135,8 @@
     addEntryIfExists(entries, expected_entries, pathname + 'empty.py?favicon', 'link');
     addEntryIfExists(entries, expected_entries, pathname + 'eventsource.py?id=eventsource', 'other');
 
+    assert_greater_than(entries.length, 0, 'There should be resrouce entries.');
```
I'm considering whether this assert should be here or inside the test_resource_entries function instead.

Any thoughts?

nhelfman replied:
Based on discussion with WebPerfWG - we should look into modifying the test to use performance observer to wait for all resources to load.

npm1 commented on resource-timing/resource_initiator_types.html:
```diff
@@ -135,6 +135,8 @@
     addEntryIfExists(entries, expected_entries, pathname + 'empty.py?favicon', 'link');
     addEntryIfExists(entries, expected_entries, pathname + 'eventsource.py?id=eventsource', 'other');
 
+    assert_greater_than(entries.length, 0, 'There should be resrouce entries.');
```
```suggestion
    assert_greater_than(entries.length, 0, 'There should be resource entries.');
```

npm1 reviewed (commented):
Looks good modulo moving the assert into a test

Commit:
Improve test reliablitity

Added performance observer to improve test reliability. Excluded eventSource type for simplicity

Commit:
add back support for event source

Since event source repeatedly triggers resource timing events we need to exclude it from the count and ensure at least one arrived.

npm1 commented on resource-timing/resource_initiator_types.html:
```diff
@@ -65,6 +66,12 @@
+function on_all_resources_loaded() {
+    all_resources_loaded = true;
+
+    step_timeout(check_finished, 100);
```
Can the timeout be a bit smaller? It seems that it should be ready shortly after it has ran this callback.

nhelfman replied:
fixed to 10ms

npm1 commented on resource-timing/resources/all_resource_types.htm:
```diff
@@ -6,6 +6,35 @@
+                const entries = list.getEntries();
+                const eventSourceCount = entries.filter(e => e.name.includes('eventsource')).length;
+
+                if(eventSource > 0) {
```
I don't follow, where is `eventSource` defined?

nhelfman replied:
fixed

npm1 commented on resource-timing/resources/all_resource_types.htm:
```diff
@@ -6,6 +6,35 @@
+                const eventSourceCount = entries.filter(e => e.name.includes('eventsource')).length;
+
+                if(eventSource > 0) {
+                    totalEventSourceCount += eventSourceCount;
```
Presumably this does not require any if, right? At worst you just add a 0.

Commit:
minor fix

npm1 commented on resource-timing/resources/all_resource_types.htm:
```diff
@@ -6,6 +6,35 @@
+                    totalEventSourceCount += eventSourceCount;
+                } 
+                
+                // exclude eventSource since it arrives repeatedly
```
Oh that's weird, what do you mean by this? The same resource arrives multiple times?

nhelfman replied:
the way it is implemented it arrives multiple times so each time it calls the observer callback

npm1 commented on resource-timing/resources/all_resource_types.htm:
```diff
@@ -6,6 +6,35 @@
+            } 
+        
+            const observer = new PerformanceObserver(perf_observer); 
+            observer.observe({entryTypes: ["resource"]});
```
There is a resource for the body background above, is that supposed to be included here or no? Seems like it could trigger some race condition. I think we should set the value below if it is meant to be included, remove the background if it's not.

nhelfman replied:
per offline discussion - moved higher to head element to reduce the race.

npm1 commented on resource-timing/resources/all_resource_types.htm:
```diff
@@ -6,6 +6,35 @@
+                if(count === 34 && totalEventSourceCount > 0) {
+                    // we are done only if all static resources arrived and at least one event source
+                    observer.disconnect();
+                    parent.on_all_resources_loaded();
```
Should this check that it has the property, like `on_test_body_created`?

nhelfman replied:
it is not needed since that is checked outside in before the final verification in `resource-timing/resource_initiator_types.html`

Commit:
remove count check

Commit:
move script to head

Commit:
reduce timeout to 10 seconds

Commit:
clean white spaces

npm1 commented:
It's possible the Firefox failure is a flakiness in the implementation itself and not about the test being flaky, so in some sense it's possible that we may want to fix the lints and then force-merge. But some investigation is required in any case.
